### ISSUE TITLE
reportGeneralTypeIssues: Pylance complains about sending any value other than None to a generator.

### ISSUE THREAD

**Issue by amirsoroush** [microsoft/pylance-release#4275]
## Environment data

-   Language Server version: v2023.4.30
-   OS and version: Pop!_OS 22.04 LTS x86_64
-   Python version (& distribution if applicable, e.g. Anaconda): 3.10.6

## Code Snippet

```python
def gen():
    for _ in range(10):
        data = yield
        print(f"Value is {data!r}")

g = gen()
next(g)
g.send(None) # Value is None
g.send(10)   # Value is 10
```

## Expected behavior

Hi,
It's fine to send any value other than `None` to generators. I would expect Pylance to be okay with that.

## Actual behavior

```none
Argument of type "Literal[10]" cannot be assigned to parameter "__value" of type "None" in function "send"
  Type cannot be assigned to type "None"Pylance[reportGeneralTypeIssues](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportGeneralTypeIssues)
```

**Comment by debonte** [comment]
This is by design. You can eliminate the error by explicitly annotating your generator function's return type. See https://github.com/microsoft/pyright/issues/4326 for more details.

```python
# I'm not sure what types you want to be able to send to gen(). Assuming ints...
def gen() -> Generator[None, int, None]:
```

**Comment by erictraut** [comment]
Yes, the type checker cannot infer the generator function's send type. As @debonte mentioned, it currently assumes it's `None`, but it would be more proper to assume `Unknown`. This would eliminate the type error. I'll make this change in pyright (the type checker upon which pylance is built).

**Comment by erictraut** [comment]
This will be addressed in the next release of pyright and included in a future release of pylance.

**Comment by debonte** [comment]
@erictraut, you mentioned in https://github.com/microsoft/pyright/issues/4326#issuecomment-1349059690 that assuming `Unknown` would be an issue for `strict` mode users. Was that incorrect?

**Comment by erictraut** [comment]
Oh, I forgot about that thread. Thanks for reminding me. The earlier bug report was about a false negative, so it was easier to dismiss. This issue is about a false positive that results from the same underlying issue.

You're right that `Unknown` could create a bunch of noise for `strict` mode users. I think we have two options if we want to eliminate the false positive:
1. Force strict-mode users to deal with the noise. This will be a one-time change, but it might be painful for some.
2. Infer `Any` rather than infer `None` or `Unknown` for the send type.

I'm slightly inclined toward option 1, but I could be convinced that option 2 is preferable. If we're going to make a change here, I'd prefer to get it right (and eliminate both the false positive and false negative cases) rather than simply mask the false positive. Let me know if you feel strongly about option 2 or have some other suggested approach.

**Comment by debonte** [comment]
I'm new to generators, but my impression is that `send` is only interesting in cases where the "return value" of `yield` is used. Could we only infer `Unknown` for unannotated generators that "use" the return value of at least one of their `yield` statements (in an assignment, passing to a function, etc)? How hard would it be to determine that?

My thinking is that any `strict` mode user that is sending values to a generator would already have had to annotate that generator. If they hadn't annotated it, with the current "assume `None`" behavior, they'd only be able to `send(None)` which is [equivalent](https://peps.python.org/pep-0342/#new-generator-method-send-value) to `next()` and therefore doesn't seem very interesting.

So if we can identify generators that were written for use with `send` and only make `SendType` `Unknown` in that case, the noise for `strict` users should be negligible. Or perhaps zero.

**Comment by erictraut** [comment]
That's an interesting idea, but I think we'd need to infer `Any` rather than `None` for it to be sound. If a generator doesn't use the "send value" (i.e. the value returned by the `yield` expression), that doesn't necessarily mean that a caller can send it only a `None` value. If we infer `Any` rather than `None`, that would be both sound (from a type perspective) and eliminate the noise.

We can use a simple heuristic to determine whether the return value of `yield` is used — namely, whether any yield expression parse node (`YieldNode`) within the function is contained within a parse node other than a `StatementListNode`. That's the same heuristic I used for the `reportUnusedExpression` check, and it seems to work well.

I just coded up this change and checked it against a couple of code bases that use generators. It did indeed eliminate all of the strict-mode noise that would otherwise result from my earlier change.
